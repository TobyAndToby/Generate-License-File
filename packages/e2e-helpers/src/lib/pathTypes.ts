import path from "path";

type DescribeCallback = (packageJsonPath: string) => void;

export const describeRelativeAndAbsolutePaths = (
  relativePackageJsonPath: string,
  callback: DescribeCallback,
) => {
  describe("for a relative path", () => {
    callback(relativePackageJsonPath);
  });

  const absolutePackageJsonPath = path.join(process.cwd(), relativePackageJsonPath);

  describe("for an absolute path", () => {
    callback(absolutePackageJsonPath);
  });
};
