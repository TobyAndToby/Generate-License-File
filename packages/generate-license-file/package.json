{
  "name": "generate-license-file",
  "version": "0.0.0",
  "description": "Generates a text file containing all of the licenses for your production dependencies",
  "main": "src/index.js",
  "license": "ISC",
  "type": "commonjs",
  "contributors": [
    {
      "name": "Toby Smith",
      "url": "https://github.com/tobysmith568"
    },
    {
      "name": "Toby Bessant",
      "url": "https://github.com/tobybessant"
    }
  ],
  "files": [
    "src",
    "bin"
  ],
  "keywords": [
    "cli",
    "license",
    "licence",
    "third",
    "party",
    "file",
    "generate",
    "oss"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/TobyAndToby/generate-license-file.git"
  },
  "bugs": {
    "url": "https://github.com/TobyAndToby/generate-license-file/issues"
  },
  "homepage": "https://generate-license-file.js.org",
  "bin": {
    "generate-license-file": "bin/generate-license-file"
  }
}
