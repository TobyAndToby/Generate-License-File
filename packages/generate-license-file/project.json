{
  "name": "generate-license-file",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "packages/generate-license-file/src",
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@nx/js:tsc",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/packages/generate-license-file",
        "main": "packages/generate-license-file/src/index.ts",
        "tsConfig": "packages/generate-license-file/tsconfig.lib.json",
        "assets": [
          "packages/generate-license-file/*.md",
          "packages/generate-license-file/bin/**",
          "LICENSE.md"
        ],
        "buildableProjectDepsInPackageJsonType": "dependencies"
      }
    },
    "run": {
      "executor": "nx:run-commands",
      "options": {
        "command": "node ./dist/packages/generate-license-file/bin/generate-license-file"
      }
    },
    "version": {
      "executor": "nx:run-commands",
      "options": {
        "command": "node tools/scripts/version.mjs generate-license-file {args.version}"
      }
    },
    "publish": {
      "executor": "nx:run-commands",
      "options": {
        "command": "node tools/scripts/publish.mjs generate-license-file {args.tag}"
      }
    },
    "lint": {
      "executor": "@nx/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["packages/generate-license-file/**/*.ts"],
        "maxWarnings": 0
      }
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "packages/generate-license-file/jest.config.ts",
        "passWithNoTests": false
      }
    }
  },
  "tags": []
}
