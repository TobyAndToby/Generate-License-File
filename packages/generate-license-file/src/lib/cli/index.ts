import { mainCommand } from "./commands/main";

export const main = async (args: string[]): Promise<void> => {
  await program.parseAsync(args);
};

const program = mainCommand;
