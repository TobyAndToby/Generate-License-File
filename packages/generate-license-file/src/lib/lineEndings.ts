import os from "os";

export const lineEndings = ["crlf", "lf"] as const;

/**
 * Used to specify which line endings to use in the generated file.
 *
 * Possible values are:
 *
 * `crlf` (`\r\n`)
 *
 * `lf` (`\n`)
 */
export type LineEnding = (typeof lineEndings)[number];

const lineEndingsMap = {
  crlf: "\r\n",
  lf: "\n",
} as const satisfies Record<LineEnding, string>;

export type LineEndingCharacters = (typeof lineEndingsMap)[LineEnding];
export const lineEndingCharacters = Object.values(lineEndingsMap);

export const isLineEnding = (input: unknown): input is LineEnding =>
  lineEndings.includes(input as LineEnding);

export const getLineEndingCharacters = (input: LineEnding | undefined): LineEndingCharacters => {
  if (input === undefined) {
    return os.EOL as LineEndingCharacters;
  }

  return lineEndingsMap[input];
};
