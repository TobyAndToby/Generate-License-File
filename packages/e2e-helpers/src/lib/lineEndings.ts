import { LineEnding } from "generate-license-file";

type DescribeEachLineEndingCallback = (lineEnding: LineEnding, lineEndingLiteral: string) => void;

const lineEndings: Record<LineEnding, string> = {
  crlf: "\r\n",
  lf: "\n",
};

export const describeEachLineEnding = (callback: DescribeEachLineEndingCallback) => {
  for (const key in lineEndings) {
    const lineEnding = key as LineEnding;
    const value = lineEndings[lineEnding];

    describe("for line ending " + key, () => callback(lineEnding, value));
  }
};
