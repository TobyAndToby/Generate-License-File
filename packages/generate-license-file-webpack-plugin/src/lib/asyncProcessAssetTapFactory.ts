import { getLicenseFileText, GetLicenseFileTextOptions } from "generate-license-file";
import { join } from "path/posix";
import { Compilation, Compiler, WebpackError } from "webpack";
import { LicenseFilePlugin } from "./licenseFilePlugin";
import { devImplementation } from "./devImplementation";
import { Options } from "./options";

export type AssetProcessingAsyncTap = (_: unknown, resolve: (error?: WebpackError) => void) => void;

const unknownError = "Unknown Error! Check for error output above.";

export const asyncProcessAssetTapFactory = (
  options: Options,
  compiler: Compiler,
  compilation: Compilation,
): AssetProcessingAsyncTap => {
  const pluginName = LicenseFilePlugin.name;

  const {
    outputFileName,
    outputFolder,
    isDev,
    pathToPackageJson,
    lineEnding,
    append,
    exclude,
    replace,
  } = options;

  const getLicenseFileTextOptions: GetLicenseFileTextOptions = {
    lineEnding,
    append,
    exclude,
    replace,
  };

  const RawSource = compiler.webpack.sources.RawSource;

  return (_, resolve) => {
    const implementation = resolveImplementation(isDev);

    implementation(pathToPackageJson, getLicenseFileTextOptions)
      .then(text => {
        const outputPath = join(outputFolder, outputFileName);
        compilation.emitAsset(outputPath, new RawSource(text));
        resolve();
      })
      .catch(error => {
        const errorMessage = `${pluginName}: ${error ?? unknownError}`;
        const webpackError = new WebpackError(errorMessage);
        compilation.errors.push(webpackError);
        resolve(webpackError);
      });
  };
};

const resolveImplementation = (isDev: boolean): typeof getLicenseFileText => {
  if (isDev) {
    return devImplementation;
  }

  return getLicenseFileText;
};
