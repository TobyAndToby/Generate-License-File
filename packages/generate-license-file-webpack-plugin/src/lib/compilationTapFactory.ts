import { Compilation, Compiler } from "webpack";
import { LicenseFilePlugin } from "./licenseFilePlugin";
import { asyncProcessAssetTapFactory } from "./asyncProcessAssetTapFactory";
import { Options } from "./options";

export type CompilationTap = (compilation: Compilation) => void;

export const compilationTapFactory = (options: Options, compiler: Compiler): CompilationTap => {
  const pluginName = LicenseFilePlugin.name;

  return compilation => {
    const processAssetTap = asyncProcessAssetTapFactory(options, compiler, compilation);

    const assetOptions = {
      name: pluginName,
      stage: Compilation.PROCESS_ASSETS_STAGE_ADDITIONAL,
    };

    compilation.hooks.processAssets.tapAsync(assetOptions, processAssetTap);
  };
};
