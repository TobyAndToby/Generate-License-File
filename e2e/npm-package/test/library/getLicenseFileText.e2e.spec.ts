import {
  describeEachLineEnding,
  describeRelativeAndAbsolutePaths,
} from "@generate-license-file/e2e-helpers";
import { getLicenseFileText, LineEnding } from "generate-license-file";
import { lineEndings } from "generate-license-file/src/lib/lineEndings";

describe("getLicenseFileText", () => {
  describeRelativeAndAbsolutePaths("./package.json", (packageJsonPath) =>
    describeEachLineEnding((lineEnding, lineEndingLiteral) => {
      let lineEndingsNotUnderTest: LineEnding[] = [];

      beforeEach(() => {
        lineEndingsNotUnderTest = lineEndings.filter((x) => x !== lineEnding);
      });

      it("should match snapshot", async () => {
        const result = await getLicenseFileText(packageJsonPath, {
          lineEnding,
        });

        expect(result).toMatchSnapshot();
      });

      it("should contain the correct line ending value", async () => {
        const expectedLineEndingValue = lineEndingLiteral;

        const result = await getLicenseFileText(packageJsonPath, {
          lineEnding,
        });

        expect(result).toContain(expectedLineEndingValue);
      });

      lineEndingsNotUnderTest.forEach((otherLineEnding) =>
        it(`should not contain the incorrect line ending value (${otherLineEnding})`, async () => {
          const incorrectLineEndingValue = lineEndingLiteral;

          const result = await getLicenseFileText(packageJsonPath, {
            lineEnding,
          });

          expect(result).not.toContain(incorrectLineEndingValue);
        })
      );
    })
  );

  it("should omit versions when omitVersions is true", async () => {
    const result = await getLicenseFileText("./package.json", {
      omitVersions: true,
    });

    expect(result).toMatchSnapshot();
  });
});
