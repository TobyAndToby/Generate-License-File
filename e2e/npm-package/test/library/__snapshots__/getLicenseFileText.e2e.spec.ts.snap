// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getLicenseFileText for a relative path for line ending crlf should match snapshot 1`] = `
"This file was generated with the generate-license-file npm package!
https://www.npmjs.com/package/generate-license-file

The following npm package may be included in this product:

 - dep-four@1.0.0

This package contains the following license:

# Dep Four

This license file is spelt \`LICENCE\`.
This license should be found.

-----------

The following npm package may be included in this product:

 - dep-one@1.0.0

This package contains the following license:

# Dep One

This license file is spelt \`LICENSE.md\`.
This license should be found.

-----------

The following npm package may be included in this product:

 - dep-seven@1.0.0

This package contains the following license:

# Dep Seven

This license should be found.

With the following notices:

This is the notice file bundled with dependency seven.

-----------

The following npm package may be included in this product:

 - dep-three@1.0.0

This package contains the following license:

# Dep Three

This license file is spelt \`LICENSE\`.
This license should be found.

-----------

The following npm packages may be included in this product:

 - dep-two-duplicate@1.0.0
 - dep-two@1.0.0

These packages each contain the following license:

# Dep Two

This license file is spelt \`LICENCE.md\`.
This license should be found.

-----------

The following npm package may be included in this product:

 - dep-five@1.0.0

This package contains the following license:

MIT

-----------

This file was generated with the generate-license-file npm package!
https://www.npmjs.com/package/generate-license-file
"
`;

exports[`getLicenseFileText for a relative path for line ending lf should match snapshot 1`] = `
"This file was generated with the generate-license-file npm package!
https://www.npmjs.com/package/generate-license-file

The following npm package may be included in this product:

 - dep-four@1.0.0

This package contains the following license:

# Dep Four

This license file is spelt \`LICENCE\`.
This license should be found.

-----------

The following npm package may be included in this product:

 - dep-one@1.0.0

This package contains the following license:

# Dep One

This license file is spelt \`LICENSE.md\`.
This license should be found.

-----------

The following npm package may be included in this product:

 - dep-seven@1.0.0

This package contains the following license:

# Dep Seven

This license should be found.

With the following notices:

This is the notice file bundled with dependency seven.

-----------

The following npm package may be included in this product:

 - dep-three@1.0.0

This package contains the following license:

# Dep Three

This license file is spelt \`LICENSE\`.
This license should be found.

-----------

The following npm packages may be included in this product:

 - dep-two-duplicate@1.0.0
 - dep-two@1.0.0

These packages each contain the following license:

# Dep Two

This license file is spelt \`LICENCE.md\`.
This license should be found.

-----------

The following npm package may be included in this product:

 - dep-five@1.0.0

This package contains the following license:

MIT

-----------

This file was generated with the generate-license-file npm package!
https://www.npmjs.com/package/generate-license-file
"
`;

exports[`getLicenseFileText for an absolute path for line ending crlf should match snapshot 1`] = `
"This file was generated with the generate-license-file npm package!
https://www.npmjs.com/package/generate-license-file

The following npm package may be included in this product:

 - dep-four@1.0.0

This package contains the following license:

# Dep Four

This license file is spelt \`LICENCE\`.
This license should be found.

-----------

The following npm package may be included in this product:

 - dep-one@1.0.0

This package contains the following license:

# Dep One

This license file is spelt \`LICENSE.md\`.
This license should be found.

-----------

The following npm package may be included in this product:

 - dep-seven@1.0.0

This package contains the following license:

# Dep Seven

This license should be found.

With the following notices:

This is the notice file bundled with dependency seven.

-----------

The following npm package may be included in this product:

 - dep-three@1.0.0

This package contains the following license:

# Dep Three

This license file is spelt \`LICENSE\`.
This license should be found.

-----------

The following npm packages may be included in this product:

 - dep-two-duplicate@1.0.0
 - dep-two@1.0.0

These packages each contain the following license:

# Dep Two

This license file is spelt \`LICENCE.md\`.
This license should be found.

-----------

The following npm package may be included in this product:

 - dep-five@1.0.0

This package contains the following license:

MIT

-----------

This file was generated with the generate-license-file npm package!
https://www.npmjs.com/package/generate-license-file
"
`;

exports[`getLicenseFileText for an absolute path for line ending lf should match snapshot 1`] = `
"This file was generated with the generate-license-file npm package!
https://www.npmjs.com/package/generate-license-file

The following npm package may be included in this product:

 - dep-four@1.0.0

This package contains the following license:

# Dep Four

This license file is spelt \`LICENCE\`.
This license should be found.

-----------

The following npm package may be included in this product:

 - dep-one@1.0.0

This package contains the following license:

# Dep One

This license file is spelt \`LICENSE.md\`.
This license should be found.

-----------

The following npm package may be included in this product:

 - dep-seven@1.0.0

This package contains the following license:

# Dep Seven

This license should be found.

With the following notices:

This is the notice file bundled with dependency seven.

-----------

The following npm package may be included in this product:

 - dep-three@1.0.0

This package contains the following license:

# Dep Three

This license file is spelt \`LICENSE\`.
This license should be found.

-----------

The following npm packages may be included in this product:

 - dep-two-duplicate@1.0.0
 - dep-two@1.0.0

These packages each contain the following license:

# Dep Two

This license file is spelt \`LICENCE.md\`.
This license should be found.

-----------

The following npm package may be included in this product:

 - dep-five@1.0.0

This package contains the following license:

MIT

-----------

This file was generated with the generate-license-file npm package!
https://www.npmjs.com/package/generate-license-file
"
`;

exports[`getLicenseFileText should omit versions when omitVersions is true 1`] = `
"This file was generated with the generate-license-file npm package!
https://www.npmjs.com/package/generate-license-file

The following npm package may be included in this product:

 - dep-four

This package contains the following license:

# Dep Four

This license file is spelt \`LICENCE\`.
This license should be found.

-----------

The following npm package may be included in this product:

 - dep-one

This package contains the following license:

# Dep One

This license file is spelt \`LICENSE.md\`.
This license should be found.

-----------

The following npm package may be included in this product:

 - dep-seven

This package contains the following license:

# Dep Seven

This license should be found.

With the following notices:

This is the notice file bundled with dependency seven.

-----------

The following npm package may be included in this product:

 - dep-three

This package contains the following license:

# Dep Three

This license file is spelt \`LICENSE\`.
This license should be found.

-----------

The following npm packages may be included in this product:

 - dep-two
 - dep-two-duplicate

These packages each contain the following license:

# Dep Two

This license file is spelt \`LICENCE.md\`.
This license should be found.

-----------

The following npm package may be included in this product:

 - dep-five

This package contains the following license:

MIT

-----------

This file was generated with the generate-license-file npm package!
https://www.npmjs.com/package/generate-license-file
"
`;
